modules/reactor/src/meshgenerators/CoreMeshGenerator.C

```diff
@@ -788,56 +816,86 @@ CoreMeshGenerator::generate()
+        // peripheral_index
+        const auto elem_rid = _drum_region_id_map[assembly_type_id][z_id][peripheral_idx];
+        elem->set_extra_integer(region_id_int, elem_rid);
+
+        // Set element block name and block id
+        auto elem_block_name = default_block_name;
+        if (getReactorParam<bool>(RGMB::region_id_as_block_name))
+          elem_block_name += "_REG" + std::to_string(elem_rid);
+        else
         {
-          bool has_background_block_name = !_background_block_name_map[assembly_type_id].empty();
-          if (has_background_block_name)
-            elem_block_name += "_" + _background_block_name_map[assembly_type_id][z_id];
+          bool has_drum_block_name = !_drum_block_name_map[assembly_type_id].empty();
+          if (has_drum_block_name)
+            elem_block_name += "_" + _drum_block_name_map[assembly_type_id][z_id][peripheral_idx];
         }
+        if (elem->type() == TRI3 || elem->type() == PRISM6)
+          elem_block_name += RGMB::TRI_BLOCK_NAME_SUFFIX;
+        updateElementBlockNameId(
+            *(*_build_mesh), elem, rgmb_name_id_map, elem_block_name, next_block_id);
+      }
+      else
+      {
+        // Element is in an assembly duct or background region since it doesn't
+        // have an assembly type id in the drum region map. Infer region id from
+        // assembly_type_id, z_id, and peripheral_index
+        bool is_background_region = peripheral_idx == 0;
+        const auto elem_rid =
+            (is_background_region
+                 ? _background_region_id_map[assembly_type_id][z_id]
+                 : _duct_region_id_map[assembly_type_id][z_id][peripheral_idx - 1]);
+        elem->set_extra_integer(region_id_int, elem_rid);
+
+        // Set element block name and block id
+        auto elem_block_name = default_block_name;
+        if (getReactorParam<bool>(RGMB::region_id_as_block_name))
+          elem_block_name += "_REG" + std::to_string(elem_rid);
         else
         {
-          bool has_duct_block_names = !_duct_block_name_map[assembly_type_id].empty();
-          if (has_duct_block_names)
-            elem_block_name +=
-                "_" + _duct_block_name_map[assembly_type_id][z_id][peripheral_idx - 1];
+          if (is_background_region)
+          {
+            bool has_background_block_name = !_background_block_name_map[assembly_type_id].empty();
+            if (has_background_block_name)
+              elem_block_name += "_" + _background_block_name_map[assembly_type_id][z_id];
+          }
+          else
+          {
+            bool has_duct_block_names = !_duct_block_name_map[assembly_type_id].empty();
+            if (has_duct_block_names)
+              elem_block_name +=
+                  "_" + _duct_block_name_map[assembly_type_id][z_id][peripheral_idx - 1];
+          }
         }
+        if (elem->type() == TRI3 || elem->type() == PRISM6)
```

Comment:
we now support second order elements, should they be added there?

Reply:
Disregarding this comment as RGMB only supports first order elements for now